File: java-workshop/part-1-powermock/src/main/java/infrastructure/DatabaseDiscountCodeRepository.java
package infrastructure;

import domain.DiscountCode;
import repository.DiscountCodeRepository;

/**
 * Repository implementation that requires a real database connection.
 *
 * This class connects to the database at construction time.
 * It will fail in test environments where no database is available.
 */
public class DatabaseDiscountCodeRepository implements DiscountCodeRepository {

    private final DatabaseConnection connection;

    public DatabaseDiscountCodeRepository() {
        // Requires real database infrastructure - will fail in tests without mocking
        this.connection = DatabaseConnection.connect();
    }

    @Override
    public boolean checkDiscountCode(String customerId, DiscountCode discountCode) {
        return connection.executeQuery(
            "SELECT used FROM discount_codes WHERE customer_id = ? AND code = ?",
            customerId, discountCode.name()
        );
    }

    @Override
    public void markAsUsed(String customerId, DiscountCode discountCode) {
        connection.executeUpdate(
            "INSERT INTO discount_codes (customer_id, code, used) VALUES (?, ?, true)",
            customerId, discountCode.name()
        );
    }
}

File: csharp-workshop/Part1.StaticDeps/Infrastructure/DatabaseDiscountCodeRepository.cs
using Part1.StaticDeps.Domain;
using Part1.StaticDeps.Repository;

namespace Part1.StaticDeps.Infrastructure;

/// <summary>
/// Repository implementation that requires a real database connection.
/// This class connects to the database at construction time.
/// It will fail in test environments where no database is available.
/// </summary>
public class DatabaseDiscountCodeRepository : IDiscountCodeRepository
{
    private readonly DatabaseConnection _connection;

    public DatabaseDiscountCodeRepository()
    {
        // Requires real database infrastructure - will fail in tests without mocking
        _connection = DatabaseConnection.Connect();
    }

    public bool CheckDiscountCode(string customerId, DiscountCode discountCode) =>
        _connection.ExecuteQuery(
            "SELECT used FROM discount_codes WHERE customer_id = @customerId AND code = @code",
            customerId, discountCode.ToString());

    public void MarkAsUsed(string customerId, DiscountCode discountCode) =>
        _connection.ExecuteUpdate(
            "INSERT INTO discount_codes (customer_id, code, used) VALUES (@customerId, @code, true)",
            customerId, discountCode.ToString());
}
